subcommand/delete-completed-job/command.go

```diff
@@ -0,0 +1,165 @@
+package deletecompletedjob
+
+import (
+	"flag"
+	"fmt"
+	"github.com/hashicorp/consul-k8s/subcommand"
+	k8sflags "github.com/hashicorp/consul-k8s/subcommand/flags"
+	"github.com/hashicorp/consul/command/flags"
+	"github.com/hashicorp/go-hclog"
+	"github.com/mitchellh/cli"
+	v1 "k8s.io/api/batch/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	"os"
+	"sync"
+	"time"
+)
+
+// Command is the command for deleting completed jobs.
+type Command struct {
+	UI cli.Ui
+
+	flags         *flag.FlagSet
+	k8s           *k8sflags.K8SFlags
+	flagNamespace string
+	flagTimeout   string
+
+	once      sync.Once
+	help      string
+	k8sClient kubernetes.Interface
+
+	// retryDuration is how often we'll retry deletion.
+	retryDuration time.Duration
+}
+
+func (c *Command) init() {
+	c.flags = flag.NewFlagSet("", flag.ContinueOnError)
+
+	c.k8s = &k8sflags.K8SFlags{}
+	c.flags.StringVar(&c.flagNamespace, "k8s-namespace", "",
+		"Name of Kubernetes namespace where the job is deployed")
+	c.flags.StringVar(&c.flagTimeout, "timeout", "30m",
+		"How long we'll wait for the job to complete before timing out, e.g. 1ms, 2s, 3m")
+	flags.Merge(c.flags, c.k8s.Flags())
+	c.help = flags.Usage(help, c.flags)
+
+	// Default retry to 30s. This is exposed for setting in tests.
+	if c.retryDuration == 0 {
+		c.retryDuration = 30 * time.Second
+	}
+}
+
+// Run will attempt to delete the job once it succeeds. If the job hits its
+// backoff limit, it will give up deleting it.
+func (c *Command) Run(args []string) int {
+	c.once.Do(c.init)
+
+	// Validate command.
+	if err := c.flags.Parse(args); err != nil {
+		return 1
+	}
+	if len(c.flags.Args()) != 1 {
+		c.UI.Error("Must have one arg: the job name to delete.")
+		return 1
+	}
+	jobName := c.flags.Args()[0]
+	if c.flagNamespace == "" {
+		c.UI.Error("Must set flag -k8s-namespace")
+		return 1
+	}
+	timeout, err := time.ParseDuration(c.flagTimeout)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("%q is not a valid timeout: %s", c.flagTimeout, err))
+		return 1
+	}
+	start := time.Now()
+
+	// c.k8sclient might already be set in a test.
+	if c.k8sClient == nil {
+		config, err := subcommand.K8SConfig(c.k8s.KubeConfig())
+		if err != nil {
+			c.UI.Error(fmt.Sprintf("Error retrieving Kubernetes auth: %s", err))
+			return 1
+		}
+
+		c.k8sClient, err = kubernetes.NewForConfig(config)
+		if err != nil {
+			c.UI.Error(fmt.Sprintf("Error initializing Kubernetes client: %s", err))
+			return 1
+		}
+	}
+
+	logger := hclog.New(&hclog.LoggerOptions{
+		Level:  hclog.Info,
+		Output: os.Stderr,
+	})
+
+	// Wait for job to complete.
+	logger.Info(fmt.Sprintf("waiting for job %q to complete successfully", jobName))
+	for {
```

Comment:
Hmm, I think I'd like to wait until we have 3 different uses before generalizing it. The use here is slightly different because the `op` that we'd pass in wouldn't need to complete without error unlike in server-acl-init. So we'd need to parameterize it in some way and I'm not sure it's worth that effort yet.

Reply:
Ok, that makes sense.